Google Actions for Assistant

philk opened the pull request:
## Description:

I'm looking for some feedback on this both as a component and as a specific implementation.

The current `emulated_hue` implementation can only work with either the Echo or Google Home, but not both at the same time so I took some of the ideas from the Alexa component.

Google provides a [Smart Home API](https://developers.google.com/actions/smarthome/create-app) for interacting with Google Assistant. It's a relatively straightforward API that just requires some mapping of Home-Assistant objects to a [couple of schemas](https://developers.google.com/actions/smarthome/light-schema). I think that part all makes sense and is a reasonable component for hass. I will acknowledge there was a lot of copy paste from the hue and Alexa components to figure out how aiohttp works. Happy to take feedback on the /right/ way to do things for sure 😅

Where I'm most looking for feedback though is really UX, basically the Google Actions deploy and setup process.  You have to [set up OAuth](https://developers.google.com/actions/smarthome/create-app#set_up_an_oauth_20_server) for your action to authenticate against and the way I did it is...hacky to say the least. I'm really curious what the path forward for the [Cloud Auth](https://github.com/home-assistant/home-assistant/tree/dev/homeassistant/components/cloud) stuff is but right now it looks pretty AWS specific and more for handling login to hass rather than being an OAuth provider.

I'm also not super happy with having to ask people to go through the [setup steps](https://developers.google.com/actions/smarthome/testing-deploying) for creating an app in GCP. We could probably break it down to:

1. Download Google's gactions tool
2. Copy a json file and fill in some fields
3. Run the `gactions test` command with the json file

Which isn't /too/ bad but doesn't feel as user friendly as hass has gotten lately (OMG, not having to deal with Zwave compilation and install, life saving).



**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3657

## Example entry for `configuration.yaml` (if applicable):
```yaml
google_assistant:
  project_id: someproject-2d0b8
  client_id: supersecretid
  access_token: extradoublesecrettoken
  exposed_domains:
    - switch
    - cover
    - light
    - group
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New files were added to `.coveragerc`.

homeassistant commented:
Hello @philk,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`37db7d3fed4efa497cd8230bbbea960eef08175d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;9632&#x2F;commits&#x2F;37db7d3fed4efa497cd8230bbbea960eef08175d) - This commit has something that looks like an email address (pkates@salesforce.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

balloob commented on .coveragerc:
```diff
@@ -74,6 +74,9 @@ omit =
     homeassistant/components/google.py
     homeassistant/components/*/google.py
 
+    homeassistant/components/googleactions.py
```
This is not needed, you wrote tests 👍

balloob commented on homeassistant/components/googleactions.py:
```diff
@@ -0,0 +1,390 @@
+    '{}/r/{}#access_token={}&token_type=bearer&state={}'
+
+
+def setup(hass: HomeAssistant, yaml_config: Dict[str, Any]):
```
Please make this async

```python
@asyncio.coroutine
def async_setup(hass, config):
```

balloob commented on homeassistant/components/googleactions.py:
```diff
@@ -0,0 +1,390 @@
+    def __init__(self, hass: HomeAssistant, cfg: Dict[str, Any]) -> None:
+        """Initialize instance of the view."""
+        super().__init__()
+        self.hass = hass
```
You can get the `hass` instance from the request, you don't have to store it in the instance:

```
# inside get(…)
hass = request.app['hass']
```

balloob commented on homeassistant/components/googleactions.py:
```diff
@@ -0,0 +1,390 @@
+        """Initialize Google Actions view."""
+        super().__init__()
+
+        self.hass = hass
```
Same.

balloob commented on homeassistant/components/googleactions.py:
```diff
@@ -0,0 +1,390 @@
+
+        return is_default_exposed or explicit_expose
+
+    def handle_sync(self, hass: HomeAssistant, request_id: str):
```
Please decouple the I/O from the handling of the messages. Have a look at the Alexa smart home skill as an example: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/alexa/smart_home.py#L32-L44

balloob replied:
I would even consider breaking it up in multiple files: `google_assistant/smart_home.py` and `google_assistant/http.py`

balloob commented on homeassistant/components/googleactions.py:
```diff
@@ -0,0 +1,390 @@
+    def post(self, request: Request):
+        """Handle Google Action requests."""
+        auth = request.headers.get('Authorization', '')
+        if self.access_token not in auth:
```
Let's do actual comparison. `not in` only checks if string A exists inside string B.

balloob reviewed (changes requested):
This is awesome and off to a great start! Yes, OAuth2 is currently pretty cumbersome but that's something we can improve later. I think for now the current flow is fine.

Added some comments, mainly around let's follow the pattern in `alexa/smart_home.py` and decouple message handling from I/O.

philk commented:
Thanks for the feedback, I hadn't looked at the rewritten Alexa component but I like that layout and I'm stealing some ideas from there that should tidy up my code quite a lot 😁 
Should just take a few days to re-org things.

Gagnon06 commented:
Hi! Great idea! I found a project that could be use as a reference : [Phlex](https://github.com/d8ahazard/phlex). The setup is somewhat easy. It is actually a server that connects to a Plex server instance to enable voice commands. It connects to Google Assistant and Alexa using a simple connection. The Google Action is officially accessible from Android phones and Google Home and is named Flex TV. This way, users don't have to create their own private Google Action but still can use a private server.

I tried this project and it seems to work properly, but I haven't look at the source since it is written in php, but I think it could be a good help for you.

Good work!

balloob commented:
We should not depend on Phlex as we should not promote the use of external gateway services.

Gagnon06 commented:
I am not talking about dependency, but more as a design reference.

philk commented:
Yeah, I'm not sure exactly how they do it but I don't want people's credentials passing through a system I run which is the only way I can think of to host one "Home-Assistant" action that shows up for everyone in their Google Home app.

arsaboo commented:
Can't wait to get this merged!!

kylerw commented:
Is there setup documentation for this, yet? I'd love to give it a go....

philk commented:
Not yet, should get some time to work on it this weekend.

isteaks commented:
Hi,

thanks for your work! very helpful feature ;)
I am not sure to understand you last email.
If my understanding is correct the Google smart home mechanism is to define
a mapping (more or less a database with a service connection) to list and
expose commands to the Google assistant cloud. so I would expect that the
"home assistant plug-in" will deal with Google assistant account
authentication and connections directly from client side. The user will
have to expose/define actions and fill his own credentials using a proper
config file...

Am I wrong?

thanks and BR,
David.

Le 9 oct. 2017 23:26, "Phil Kates" <notifications@github.com> a écrit :

> Yeah, I'm not sure exactly how they do it but I don't want people's
> credentials passing through a system I run which is the only way I can
> think of to host one "Home-Assistant" action that shows up for everyone in
> their Google Home app.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/9632#issuecomment-335295126>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AeTwicvIxmitMI-r5XNTfBxrN9IFxynGks5sqo96gaJpZM4Ppswc>
> .
>

philk commented:
The way the Google Actions works is either through a service like API.ai (now DialogFlow) or through the Google Compute Platform (via the Actions API). The short version (since I'm currently working on the step by step docs) is:

1. [Create a project](https://developers.google.com/actions/sdk/create-a-project) in the Google Dev Console
2. Fill out all the information for the project like URLs, name, etc and get your OAuth credentials.
3. Put your project in "Test" (exposes it to your account)
4. Add config keys (project_id, client_id, access_token) to Hass `configuration.yaml`
5. Add your app/project in the Google Home app

Like I said in my initial comment it's a lot more complicated than I'd like it to be but at least somewhat less so than the current downgrade the Google Home app process to get emulated_hue working.

cnrd commented:
> than the current downgrade the Google Home app process to get emulated_hue working.

I know it probably isn't of much use, but it looks like Google (or Philips) blocked this from working, as even the downgraded app now redirect to the Hue auth page. (I haven't tried downgrading the Google app or anything like that yet, to see if that could fix it).

roofuskit commented:
Emulated Hue will not work at all now.

isteaks commented:
Philips just change a little bit it's protocol to break the compatibility
with the new Ikea connected lights as both are using zigbee... that company
that volontuary are not compliant we the standard....
I suppose this is why th Haas implementation is broken..

Le 15 oct. 2017 07:23, "Conrad Juhl Andersen" <notifications@github.com> a
écrit :

> than the current downgrade the Google Home app process to get emulated_hue
> working.
>
> I know it probably isn't of much use, but it looks like Google (or
> Philips) blocked this from working, as even the downgraded app now redirect
> to the Hue auth page. (I haven't tried downgrading the Google app or
> anything like that yet, to see if that could fix it).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/9632#issuecomment-336671534>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AeTwiTHqqCT3s9so-BZdCLhvRMIgmdGPks5ssTRTgaJpZM4Ppswc>
> .
>

kylerw commented:
@philk Sorry to be impatient.... But, any update on the setup guide so some of us can test it?

kylerw commented:
attempted to set this up and got a good way there. During the setup process within the Home app i'm getting:

`
Traceback (most recent call last):
File "/usr/local/lib/python3.6/site-packages/aiohttp/web_protocol.py", line 422, in start
resp = yield from self._request_handler(request)
File "/usr/local/lib/python3.6/site-packages/aiohttp/web.py", line 306, in _handle
resp = yield from handler(request)
File "/usr/local/lib/python3.6/asyncio/coroutines.py", line 213, in coro
res = yield from res
File "/usr/local/lib/python3.6/asyncio/coroutines.py", line 213, in coro
res = yield from res
File "/config/custom_components/http/ban.py", line 58, in ban_middleware_handler
return (yield from handler(request))
File "/usr/src/app/homeassistant/components/http/__init__.py", line 428, in handle
result = yield from result
File "/usr/local/lib/python3.6/asyncio/coroutines.py", line 210, in coro
res = func(*args, **kw)
File "/config/custom_components/googleactions.py", line 139, in get
headers={'Location': generated_url})
TypeError: json_message() got an unexpected keyword argument 'headers'[0m
`

kylerw commented:
Looks like it didn't like the __init__.py in the custom_components/http folder so I overwrote it. Now I'm getting:

`
Couldn't update setting. Check your connection.
`

Within the Home App after it redirects.

cnrd commented:
I get something close to that, it redirects in the Google Home app, goes to a spinner, then says: "Something went wrong, please try again".

philk commented:
Wow, didn't realize people wanted this so badly 😅

Here's a [WIP setup doc](https://gist.github.com/philk/fb7ecd97db0915ec5748daba479dbbb4), you can also ping me on the hass discord (same name) and I'll see if I can help you out. As is this does at least _work_ if you're willing to hack it a bit since I've been using it for about a month now.

As for progress I've gotten a little bogged down in trying to clean up the tests (required because of how much I've moved around) but making progress and on the good news front stealing ideas from the alexa component will make it easier to add spectrum and color temperature support (and eventually thermostat in a future PR).

cnrd commented:
@philk thanks for WIP guide, now it goes through the auth process, but unfortunately it just kicks me back to the "Home Control" screen without the "app" setup.

I have sent you a couple of screenshots and a video of the behavior in a PM on Discord, I really hope that you have a minute to look at those and see if there is anything wrong with my setup :-)

People want it really badly now, as emulated_hue has stopped working for the Google Home, so this is currently the only thing that looks like it could get the Google Home to control stuff again ;-)

r4nd0mbr1ck commented:
@cnrd : I had that happen when I was trying to build the integration myself a while back. In my case, it had something to do with the OAuth flow but didn't have time to troubleshoot.

kylerw commented:
Works like a charm for me!!! A note, your guide says api/Google_assistant
but the .py file is api/Googleactions

On Oct 16, 2017 7:33 PM, "Conrad Juhl Andersen" <notifications@github.com>
wrote:

> @philk <https://github.com/philk> thanks for WIP guide, now it goes
> through the auth process, but unfortunately it just kicks me back to the
> "Home Control" screen without the "app" setup.
>
> I have sent you a couple of screenshots and a video of the behavior in a
> PM on Discord, I really hope that you have a minute to look at those and
> see if there is anything wrong with my setup :-)
>
> People want it really badly now, as emulated_hue has stopped working for
> the Google Home, so this is currently the only thing that looks like it
> could get the Google Home to control stuff again ;-)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/9632#issuecomment-337091062>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWorCXAFOn4pTXDpibz8rbnYEVN4IOiks5stAPtgaJpZM4Ppswc>
> .
>

cnrd commented:
Are you guys on iOS or Android?

kylerw commented:
Android

On Oct 16, 2017 7:45 PM, "Conrad Juhl Andersen" <notifications@github.com>
wrote:

> Are you guys on iOS or Android?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/9632#issuecomment-337092749>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWorF1v7MRR9giRXAq4q8kI5wPrElRXks5stAbAgaJpZM4Ppswc>
> .
>

cnrd commented:
We got it fixed on Discord, it was a problem with OAuth not liking special characters.

philk commented:
Ok, this is now new and improved. I did it on a new commit to make it more obvious but I can rebase out the googleactions commit once we're ready to merge (or do it now if that's what you guys want). Needs some testing if @cnrd and @kylerw want to take a look at the new version (sorry I got this done right after you got the original set up)

Still todo (preferably in future PRs):
- Color spectrum and temperature
- Thermostats

333ryan18 commented:
I got the new commit setup and I am having the same troubles that @kylerw was having earlier. Not sure how he solved it. BTW @philk you are going to make many people happy when this enters home assistant, many thanks.

> Couldn't update setting. Check your connection.

balloob commented on homeassistant/components/google_assistant/http.py:
```diff
@@ -0,0 +1,179 @@
+            res = yield from self.handle_execute(hass, request_id,
+                                                 payload.get('commands', []))
+
+        if res:
```
this code will raise an exception if none of the if statements was true as `res` has never been defined:

![image](https://user-images.githubusercontent.com/1444314/31647612-1a135636-b2be-11e7-8a23-84a05fd71a43.png)

philk replied:
Yeah...that was just a brainfart. I set `res = None` just above that section now.

balloob commented on homeassistant/components/google_assistant/http.py:
```diff
@@ -0,0 +1,179 @@
+            make_actions_response(request_id, {'devices': devices}))
+
+    @asyncio.coroutine
+    def handle_execute(self,
```
Interesting, I would have expected these methods too to be part of `smart_home.py`, just passing in `(config, hass, command)`. Anyway, I know you've put a lot of time in it already so I think it's fine as it is 👍

philk replied:
Yup, I think there's some more refactoring on where everything lives. I agree there's still too much logic happening in this view. I need to level up my pytest and aiohttp skills a bit first I think ;)

balloob commented on .coveragerc:
```diff
@@ -74,6 +74,9 @@ omit =
     homeassistant/components/*/google.py
 
+    homeassistant/components/googleactions.py
+    homeassistant/components/*/googleactions.py
```
Please remove these lines

balloob commented:
Very few changes needed to get this merged !

kylerw commented:
@philk so far so good! I like the changes! looks like fan domain got stripped. Plans to bring that back!? 

@333ryan18 for me it was a wrong URL in the gactions Jason and adding the permissions for the app.

philk committed:
http: Add headers key to json[_message]

philk committed:
Add google_assistant component

This component provides API endpoints for the Actions on Google Smart
Home API to interact with Google Assistant.

philk committed:
google_assistant: Re-add fan support

philk commented:
Assuming this build passes which is should I think this is good to go with all feedback addressed. Would love any feedback especially from those that set this up on the [doc PR](https://github.com/home-assistant/home-assistant.github.io/pull/3657)

@kylerw re-adding the fan domain was pretty simple I think but I also don't have a real fan to test it with. The demo platform fan looks right though and the logic for execute makes sense for fan. Only supports on/off though, looks like fans do speed low/medium/high which doesn't map as seamlessly to covers as brightness does.

cnrd commented:
Not sure if you wanted it, but did you test urlencode on the return? If it works for special characters in that case, I really think you should put it in, as it would lower the amount of support needed whenever the platform is implemented :-)

333ryan18 commented:
@kylerw Just got it working, turns out changing both the client_id and the access_token did the trick.

kylerw commented:
Fans are usually 33,66,99 for low, medium, high but most components can be
customized. Could you default to those values and allow customizing the fan
domain.

include:
fan:
- low: 36
- med: 67
- high: 92

On Oct 17, 2017 2:32 AM, "Phil Kates" <notifications@github.com> wrote:

> Assuming this build passes which is should I think this is good to go with
> all feedback addressed. Would love any feedback especially from those that
> set this up on the doc PR
> <https://github.com/home-assistant/home-assistant.github.io/pull/3657>
>
> @kylerw <https://github.com/kylerw> re-adding the fan domain was pretty
> simple I think but I also don't have a real fan to test it with. The demo
> platform fan looks right though and the logic for execute makes sense for
> fan. Only supports on/off though, looks like fans do speed low/medium/high
> which doesn't map as seamlessly to covers as brightness does.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/9632#issuecomment-337157962>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWorA6qxJVAJrggFCoiz7AnMorHjQAHks5stGYGgaJpZM4Ppswc>
> .
>

cnrd commented:
The customization of fan speed should probably be in customize.yaml, so that it can be set differently for each fan.

kylerw commented:
@cnrd yeah, it's already there for domains that support or need the customization (mqtt). For Zwave, it appears to be hardcoded. If the Google Assistant component can get those speed settings so that it can translate from a Percentage (since we are emulating a light) to the speed, then even better. The Zwave Fan Entity does have a 'speed_list' but the numeric values are not available directly.

philk committed:
google_assistant: Fix Scene handling

- The way I originally wrote the MAPPING_COMPONENT and the way it's actual
  used changed so the comment was updated to match that.
- Use const's in more places
- Handle the ActivateScene command correctly

philk committed:
google_assistant: Fix flakey compare test

Was failing on 3.4.2 and 3.5, this is more correct anyway.

cytech commented:
trying an install to test follwing latest doc
when i go to Home app and add the device i am receiving the same error as @kylerw  21 hours ago.
Home app shows: 500 internal server error. Server got itself in trouble.
and
Hass log shows:
2017-10-17 12:34:02 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.5/site-packages/aiohttp/web_protocol.py", line 422, in start
    resp = yield from self._request_handler(request)
  File "/srv/homeassistant/lib/python3.5/site-packages/aiohttp/web.py", line 306, in _handle
    resp = yield from handler(request)
  File "/usr/lib/python3.5/asyncio/coroutines.py", line 213, in coro
    res = yield from res
  File "/usr/lib/python3.5/asyncio/coroutines.py", line 213, in coro
    res = yield from res
  File "/home/homeassistant/.homeassistant/custom_components/http/ban.py", line 58, in ban_middleware_handler
    return (yield from handler(request))
  File "/srv/homeassistant/lib/python3.5/site-packages/homeassistant/components/http/__init__.py", line 428, in handle
    result = yield from result
  File "/usr/lib/python3.5/asyncio/coroutines.py", line 210, in coro
    res = func(*args, **kw)
  File "/home/homeassistant/.homeassistant/custom_components/google_assistant/auth.py", line 80, in get
    headers={'Location': generated_url})
TypeError: json_message() got an unexpected keyword argument 'headers'
@kylerw  What did you do to correct this?
thanx
Dave

cnrd commented:
@cytech You need to add most of the files in homeassistant/components/google_assistant and replace one of the __init__.py (The second one in files changed) to homeassistant/components/http. The component won't work as a custom_component

cytech commented:
@cnrd Ok Ill give that a shot.
I did have them under custom_components
thanx

Z-pi commented:
Have you seen the new device types? [](https://developers.google.com/actions/smarthome/guides/)

cytech commented:
@cnrd ok no errors in hass now
but now ithe Home app attempts to link the device and returns
Couldn't update setting. Check your connection.
i saw earlier you had a similar issue?

cnrd commented:
@cytech Make sure the client_id and access_token does not contain special chars (Only use alphanumeric characters), try making the client_id less than 20 chars long, and the access_token less than 80. Also make sure that the url in your projects.yaml is correct and that you are using a HTTPS connection that is not using a selfsigned cert (Use a Let's Encrypt cert, it's free).

cytech commented:
@cnrd  ok thanx
all alphanumeric
client id = 20 chars   token 54
but i am using  a self signed certificate that I just put on there for this component.
I'll try to go the full route.
thanx

arsaboo commented:
@all Guys, let us use [this thread](https://community.home-assistant.io/t/setting-up-google-actions-for-assistant/29889) in the forums to discuss support/configuration issues and not pollute this PR. 

Others will face similar issues and would benefit from this discussion.

philk committed:
google_assistant: Use volume attr for media_player

kylerw commented:
I'm getting "alias must be a list" repeated in the logs.

Also, can you add brightness to the fan domain? With emulated hue I was then able to say "set the fan to 33 percent" which would set it to low. Using Google Home app I could then set a shortcut that was "set the fan to low" which translates to the above statement. Not ideal but worked well.

kylerw commented:
@philk - is this doing a SYNC to update the available devices (add new or remove those that don't exist)? In my testing, it doesn't appear to be - it would be nice so as new devices are added we don't have to unlink/relink and reassign rooms.

arsaboo commented:
@kylerw Please don't use the PR for support/feature requests.

balloob commented:
Deleting all comments not related to the code.

balloob commented:
🎉  🐬  Thanks.
